src/ray/raylet/lineage_cache.cc

```diff
@@ -198,106 +183,52 @@ LineageCache::LineageCache(const ClientID &client_id,
 
 /// A helper function to add some uncommitted lineage to the local cache.
 void LineageCache::AddUncommittedLineage(const TaskID &task_id,
-                                         const Lineage &uncommitted_lineage,
-                                         std::unordered_set<TaskID> &subscribe_tasks) {
+                                         const Lineage &uncommitted_lineage) {
+  RAY_LOG(DEBUG) << "Adding uncommitted task " << task_id << " on " << client_id_;
   // If the entry is not found in the lineage to merge, then we stop since
   // there is nothing to copy into the merged lineage.
   auto entry = uncommitted_lineage.GetEntry(task_id);
   if (!entry) {
     return;
   }
-  RAY_CHECK(entry->GetStatus() == GcsStatus::UNCOMMITTED_REMOTE);
+  RAY_CHECK(entry->GetStatus() == GcsStatus::UNCOMMITTED);
 
   // Insert a copy of the entry into our cache.
   const auto &parent_ids = entry->GetParentTaskIds();
   // If the insert is successful, then continue the DFS. The insert will fail
   // if the new entry has an equal or lower GCS status than the current entry
   // in our cache. This also prevents us from traversing the same node twice.
   if (lineage_.SetEntry(entry->TaskData(), entry->GetStatus())) {
-    subscribe_tasks.insert(task_id);
+    // TODO(swang): Do not subscribe to tasks that are in the uncommitted
+    // lineage but were already COMMITTED at the sender.
+    RAY_CHECK(SubscribeTask(task_id));
```

Comment:
So the plan is for this to go away and to instead flush all lineage caches in the event of a failure, right?

Reply:
No, I think that would be in addition to this PR, not instead of...unless I'm misunderstanding what you mean by "this"?

Reply:
Oh, I was thinking the subscription would go away, but I agree now that it should stay.